include/sparrow/buffer.hpp

```diff
@@ -18,454 +18,870 @@
+    constexpr auto buffer<T>::rend() noexcept -> reverse_iterator
     {
         return reverse_iterator(begin());
     }
 
     template <class T>
-    auto buffer_base<T>::rbegin() const -> const_reverse_iterator
+    constexpr auto buffer<T>::rbegin() const noexcept -> const_reverse_iterator
     {
         return crbegin();
     }
-
+    
     template <class T>
-    auto buffer_base<T>::rend() const -> const_reverse_iterator
+    constexpr auto buffer<T>::rend() const noexcept -> const_reverse_iterator
     {
         return crend();
     }
 
     template <class T>
-    auto buffer_base<T>::crbegin() const -> const_reverse_iterator
+    constexpr auto buffer<T>::crbegin() const noexcept -> const_reverse_iterator
     {
-        return const_reverse_iterator(cend());
+        return const_reverse_iterator(end());
     }
 
     template <class T>
-    auto buffer_base<T>::crend() const -> const_reverse_iterator
+    constexpr auto buffer<T>::crend() const noexcept -> const_reverse_iterator
     {
-        return const_reverse_iterator(cbegin());
+        return const_reverse_iterator(begin());
     }
 
     template <class T>
-    void buffer_base<T>::swap(buffer_base<T>& rhs) noexcept
+    constexpr bool buffer<T>::empty() const noexcept
     {
-        std::swap(p_data, rhs.p_data);
-        std::swap(m_size, rhs.m_size);
+        return get_data().p_begin == get_data().p_end;
     }
-
+    
     template <class T>
-    bool buffer_base<T>::equal(const buffer_base<T>& rhs) const
+    constexpr auto buffer<T>::capacity() const noexcept -> size_type
     {
-        return m_size == rhs.m_size && std::equal(p_data, p_data + m_size, rhs.p_data);
+        return static_cast<size_type>(get_data().p_storage_end - get_data().p_begin);
     }
 
     template <class T>
-    bool operator==(const buffer_base<T>& lhs, const buffer_base<T>& rhs)
+    constexpr auto buffer<T>::size() const noexcept -> size_type
     {
-        return lhs.equal(rhs);
+        return static_cast<size_type>(get_data().p_end - get_data().p_begin);
     }
 
-    /*************************
-     * buffer implementation *
-     *************************/
-
     template <class T>
-    buffer<T>::buffer(size_type size)
-        : base_type{allocate(size), size}
+    constexpr auto buffer<T>::max_size() const noexcept -> size_type
     {
+        return max_size_impl(get_allocator());
     }
-
+    
     template <class T>
-    buffer<T>::buffer(size_type size, value_type value)
-        : base_type{allocate(size), size}
+    constexpr void buffer<T>::clear()
     {
-        std::fill_n(base_type::data(), size, value);
+        erase_at_end(get_data().p_begin);
     }
 
     template <class T>
-    buffer<T>::buffer(pointer data, size_type size)
-        : base_type{data, size}
+    constexpr void buffer<T>::reserve(size_type new_cap)
     {
+        if (new_cap > max_size())
+        {
+            throw std::length_error("buffer::reserve called with new_cap > max_size()");
```

Comment:
What should the policy for raising exception be?

Reply:
I took the same as `std::vector`, see the `Exceptions` section [here](https://en.cppreference.com/w/cpp/container/vector/vector) and [here](https://en.cppreference.com/w/cpp/container/vector/reserve)